src/movement_controls.py

```diff
@@ -0,0 +1,170 @@
+
+# This is setup for the Player to call "drive_input" where the vehicle updates it's
+# intended velocity and rotation based on the player input
+# then the player calls "movement" to act on the change from "drive_input"
+# this way the inputs are always registered but the movement can be skipped
+# if a collision or game state makes it where the vehicle shouldnt act
+# This will help avoid race conditions on collisions
+#
+#
+# for checking if a vehicle should take damage from an impact. as a percentage of their top acceleration
+
+DEFAULT_WORLD_SIZE = 750
+DEFAULT_ACCELERATION_RATE = 2.5
+DEFAULT_BRAKE_RATE = 0.8
+DEFAULT_FRICTION = 0.1
+DEFAULT_DRIVE_SPEED = 150
+DEFAULT_TURN_SPEED = 75
+
+IMPACT_FORCE_FLOOR = 1.5
+
+
+class MovementControls:
+
+    # The stat for max (x,y) linear speed
+    drive_speed: float
+    # The stat for max rotational speed
+    turn_speed: float
+
+    # how fast the current acceleration goes from 0 -> 1
+    acceleration_rate: float
+    # how fast the brake applies to current acceleration
+    brake_rate: float
+    # how fast the vehicle's current acceleration slows while not accelerating
+    friction: float
+
+    # These are the values the vehicle updates in "drive_input"
+    current_velocity_x: float
+    current_velocity_y: float
+    current_velocity_turn: float
+    current_acceleration: float
+
+    def __init__(self, sprite: arcade.Sprite):
+        self.shadow_sprite = sprite
+        self.shadow_sprite.center_x = sprite.center_x
+        self.shadow_sprite.center_y = sprite.center_y
+        self.shadow_sprite.angle = sprite.angle
+        self.debug_world_boundary = DEFAULT_WORLD_SIZE
+        self.current_velocity_x = 0
+        self.current_velocity_y = 0
+        self.current_velocity_turn = 0
+        self.current_acceleration = 0
+        self.external_velocity_turn = 0
+
+        self.acceleration_rate = DEFAULT_ACCELERATION_RATE
+        self.brake_rate = DEFAULT_BRAKE_RATE
+        self.friction = DEFAULT_FRICTION
+        self.drive_speed = DEFAULT_DRIVE_SPEED
+        self.turn_speed = DEFAULT_TURN_SPEED
+        self.impact_buffer = 0
+
+    #   Drive input is called by the player and passed that player's input
+    #   to set what the vehicle's velocity and rotation should be
+    #   Does not change the vehicles position
+    def drive_input(
+        self, delta_time: float, input: PlayerInput, vehicle: arcade.Sprite
+    ):
+
+        if self.impact_buffer > 0:
+            self.impact_buffer -= 1
+            if self.impact_buffer <= 0:
+                self.external_velocity_turn = 0
+
+        # updates the vehicles intended  velocity and rotation based on type
+        acceleration_change = 0
+        self.shadow_sprite.center_x = vehicle.center_x
+        self.shadow_sprite.center_y = vehicle.center_y
+
+        if input.accelerate_axis.value:
+            if self.current_acceleration < 0:
+                acceleration_change += self.brake_rate
```

Comment:
I'm not following the logic here.  If current_acceleration < 0, but the player is slightly applying the gas while firmly applying the brake, then the brake rate is added to the acceleration, thereby negating the braking?  I'm probably missing something here

Reply:
This is because over a month ago when this was ready to merge accelerate and brake were 'buttons' and someone changed them to an 'axis'